README.md
# facialRecog-OpenCV

**Objetivo:** Lograr crear un sistema de seguridad, el cual cuente con reconocimiento facial como método de desbloqueo. Va a contar con una base de datos en la cual se van a registrar los usuarios a los cuales se les permite la entrada. A los que no, luego de haber intentado 3 veces de reconocer su cara y todavía nada, se va a mandar una notificación a algún celular definido por el usuario, diciendo que hay una persona no identificada intentando entrar en el dominio. Una vez logrado el objetivo, se proseguirá con la creación de un sitio web donde se puede ver el paso a paso de cómo ha sido creado, y con las explicaciones necesarias para que otra persona pueda replicarlo.

**Materiales necesarios**
  - Raspberry Pi 3B o 3B+ (cualquiera de las 2 sirve).
  - Alguna cámara web que no utilicemos, o en caso de querer comprar una, con que tenga una resolución mínima de 1280x720 perfecto.
  - Memoria SD de al menos 32gb clase 10 (donde se va a guardar el sistema operativo de la Raspberry Pi).

# Python y OpenCV!
Se va a trabajar con Python como lenguaje de programación, ya que es bastante simple, que cualquiera puede aprender a utilizar en cuestión de pocas horas.
También vamos a estar trabajando con una librería muy popular llamada “OpenCV”, la cual es utilizada para proyectos realizados con la visión de la computadora.


changelog.txt
8/06
- Creacion de speechRecognition.py y se empieza a querer meter un sistema de reconocimiento de voz para las personas que tengan ciertas dificultades.

13/06
- Actualizacion de speechRecognition_mic.py a una version estable y medianamente funcional, donde se manejan los posibles errores que pueden llegar a aparecer
- Creacion de un juego de adivinar la palabra utilizando la voz V1.0

18/06
- Instalacion de paquete Kivy, para la creacion de una GUI.
- Se intenta utilizar un motor que funcione por streaming para asi agilizar el proceso de reconocimiento de voz. Por el momento se analiza pocketsphinx. (El archivo test.py intenta hacerlo)

22/06
- Instalacion de PocketSphinx como motor de voz. Intento que sea completamente offline.
https://github.com/cmusphinx/pocketsphinx-python


face-test.py
import numpy as np
import cv2 as cv

face_cascade = cv.CascadeClassifier('data/haarcascade_frontalface_default.xml')
eye_cascade = cv.CascadeClassifier('data/haarcascade_eye.xml')

img = cv.imread('sachin.jpg')
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

faces = face_cascade.detectMultiScale(gray, 1.3, 5)
for (x,y,w,h) in faces:
    cv.rectangle(img,(x,y),(x+w,y+h),(255,0,0),2)
    roi_gray = gray[y:y+h, x:x+w]
    roi_color = img[y:y+h, x:x+w]
    eyes = eye_cascade.detectMultiScale(roi_gray)
    for (ex,ey,ew,eh) in eyes:
        cv.rectangle(roi_color,(ex,ey),(ex+ew,ey+eh),(0,255,0),2)
cv.imshow('img',img)
cv.waitKey(0)
cv.destroyAllWindows()


tests/pocket.py
import os
from pocketsphinx import LiveSpeech, get_model_path

model_path = get_model_path()

speech = LiveSpeech(
    verbose=False,
    sampling_rate=16000,
    buffer_size=2048,
    no_search=False,
    full_utt=False,
    hmm= os.path.join(model_path, 'es-es'),
    lm= os.path.join(model_path, 'es-20k.lm.bin'),
    dict= os.path.join(model_path, 'es.dict')
)
for phrase in speech:
    print(phrase)


ipcamera-test.py
import urllib
import cv2
import numpy as np
import time

#Importando la url
url = "http://192.168.1.37:8080/"

while True:

    #Usar urlLib para capturar la imagen y convertirla en un formato legible para cv2

    imgResp = urllib.urlopen
    (url)
    imgNp = np.array(bytearray(imgResp.read()), dtype=np.uint8)
    img = cv2.imdecode(imgNp, -1)

    #Mostrar imagen
    cv2.imshow("Camara del celular", img)

    #Darle menos presion al procesador
    #time.sleep(0.1)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break


game.py
# Juego de adivinar la palabara utilizando la voz

import speech_recognition as sr
import random
import time

# Inicializando el microfono
r = sr.Recognizer()
mic = sr.Microphone()
print("Cargando lo necesario...")


#Creando la lista de palabras a utilizar
palabras = ["apple", "orange", "banana", "watermelon", "kiwi"]
palabra_random = palabras[random.randint(0,5)]
print("La palabra aleatoria se ha generado! Intente adivinar si es Apple, Orange, Banana, Watermelon o kiwis.")
print("Buena suerte!")

while True:
    with mic as source:
        print("/nHable")
        audio = r.listen(source)

    try:
        texto = r.recognize_google(audio, language="en-EN")
        print(texto)

        if texto == palabra_random:
            print("Felicitaciones! Haz acertado!!")
            time.sleep(2)
            print("Ahora se procedera a cerrar el programa")
            break
        elif texto == "stop":
            print("Haz dicho stop, por lo que se cerrara el programa. Gracias por jugar!")
            print("Cerrando...")
        elif texto == "what's the word":
            print(palabra_random)
        else:
            print("Palabra equivocada! Sigue intentando")
    except sr.UnknownValueError:
        print("No se ha entendido lo que haz dicho, intentalo de nuevo!")

    except sr.RequestError:
        print("No dispones de una conexion a internet. Intentalo de nuevo cuando estes conectado!")
        print("Cerrando...")
        break




faces-train.py
import cv2
import os
import numpy as np
from PIL import Image
import pickle

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
image_dir = os.path.join(BASE_DIR, "images")

face_cascade = cv2.CascadeClassifier("data/haarcascade_frontalface_alt2.xml")
recognizer = cv2.face.LBPHFaceRecognizer_create()

current_id = 0
label_ids = {}
x_train = []
y_labels = []

for root, dirs, files in os.walk(image_dir):
    for file in files:
        if file.endswith("png") or file.endswith("jpg"):
            path = os.path.join(root, file)
            label = os.path.basename(root).replace(" ", "-").lower()
            #print(label, path)
            if not label in label_ids:
                label_ids[label] = current_id
                current_id += 1
            id_ = label_ids[label]
            # print(label_ids)

            pil_image = Image.open(path).convert("L")  # Convierte a gris
            size = (550, 550)
            final_image = pil_image.resize(size, Image.ANTIALIAS)
            image_array = np.array(final_image, "uint8")  # Convierte a un array de numeros la img gris

            faces = face_cascade.detectMultiScale(image_array, scaleFactor=1.5, minNeighbors=5)

            for (x, y, w, h) in faces:
                roi = image_array[y:y + h, x:x + w]
                x_train.append(roi)
                y_labels.append(id_)

# print(y_labels)
# print(x_train)

with open("labels.pickle", "wb") as f:
    pickle.dump(label_ids, f)

recognizer.train(x_train, np.array(y_labels))
recognizer.save("trainner.yml")


face_recognition.py
import face_recognition
import cv2
import numpy as np

# Empezar a capturar video de la camara web 0 (la que viene)
video_capture = cv2.VideoCapture(0)


# Cargue otra imagen de entrenamiento y que aprenda a reconocerla
chad = face_recognition.load_image_file('./data_img/chad.jpeg')
chad_encoding = face_recognition.face_encodings(chad)[0]

# Cargue otra imagen de entrenamiento y que aprenda a reconocerla
will = face_recognition.load_image_file('./data_img/will.jpeg')
will_encoding = face_recognition.face_encodings(will)[0]

# Cargue otra imagen de entrenamiento y que aprenda a reconocerla
marcos = face_recognition.load_image_file('./data_img/marcos.jpeg')
marcos_encoding = face_recognition.face_encodings(marcos)[0]

# Cargue otra imagen de entrenamiento y que aprenda a reconocerla
valen = face_recognition.load_image_file('./data_img/valen.jpeg')
valen_encoding = face_recognition.face_encodings(valen)[0]


# Crear una array con los encodings de las caras y sus nombres
known_face_encodings = [
    agus_encoding,
    chad_encoding,
    will_encoding,
    marcos_encoding,
    valen_encoding

]
known_face_names = [
    "Agus",
    "Chad",
    "Will",
    "Marcos",
    "Valen"
]

# Inicializar algunas variables (aunque no es muy necesario, lo hago para ver que variables voy a usar)
face_locations = []
face_encodings = []
face_names = []
process_this_frame = True

while True:
    # Toma un solo frame de video
    ret, frame = video_capture.read()

    # Cambia el tamaño del frame a 1/4 para que el procesamiento sea mas rapido
    small_frame = cv2.resize(frame, (0, 0), fx=0.5, fy=0.5)

    # Convierte la imagen de un perfil BGR a RGB (que face_recognition usa)
    rgb_small_frame = small_frame[:, :, ::-1]

    # Solo procesa cada otro frame del video para ahorrar tiempo
    if process_this_frame:
        # Encuentra todas las caras y los encodings del frame actual del video
        face_locations = face_recognition.face_locations(rgb_small_frame)
        face_encodings = face_recognition.face_encodings(rgb_small_frame, face_locations)

        face_names = []
        for face_encoding in face_encodings:
            # Mira si el rostro hace match con alguna de los rostros conocidos, sino le da el nombre de "Desconocido"
            matches = face_recognition.compare_faces(known_face_encodings, face_encoding)
            name = "Desconocido"

            # Usa el rostro conocico y le da el nombre correspondiente
            face_distances = face_recognition.face_distance(known_face_encodings, face_encoding)
            best_match_index = np.argmin(face_distances)

            # Funcion IF donde se le aplica el accionamiento al reconocer el rostro
            if matches[best_match_index]:
                name = known_face_names[best_match_index]

            face_names.append(name)

    process_this_frame = not process_this_frame

    # Que muestre los resultados
    for (top, right, bottom, left), name in zip(face_locations, face_names):
        # Volver a resolucion original el frame, ya que el que detectamos estaba a 1/4
        top *= 2
        right *= 2
        bottom *= 2
        left *= 2

        # Dibujar una caja alrededor de la cara
        cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 255), 2)

        # Dibujar un campo de texto con el nombre de la persona abajo
        cv2.rectangle(frame, (left, bottom - 35), (right, bottom), (0, 0, 145), cv2.FILLED)
        font = cv2.FONT_HERSHEY_DUPLEX
        cv2.putText(frame, name, (left + 6, bottom - 6), font, 0.7, (255, 255, 255), 1)

    # Que muestre la imagen final
    cv2.imshow('Video', frame)

    # Apretar Q para salir
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

video_capture.release()
cv2.destroyAllWindows()


speechRecognition_mic.py
# Archivo encargado del reconocimiento de voz V1.0 || Solo funciona en ingles por el momento

import speech_recognition as sr

r = sr.Recognizer()

# Creando la instancia de microfono
mic = sr.Microphone()

while True:
    with mic as source:
        # Quitar comentario de la linea de abajo cuando se este trabajando en un ambiente muy ruidoso
        #r.adjust_for_ambient_noise(source)
        print("Hable")
        audio = r.listen(source)

    try:
        texto = r.recognize_google(audio, language="en-EN")
        print("Dijiste: " + texto)
        if texto == "stop":
            print("Cerrando...")
            break
    except sr.UnknownValueError:
        print("No se entendio")
    except sr.RequestError:
        print("No se pudo acceder a una conexion de internet... Verifica tu conexion e intentalo de nuevo!")


faces.py
import numpy as np
import cv2
import pickle

face_cascade = cv2.CascadeClassifier("data/haarcascade_frontalface_alt2.xml")
eye_cascade = cv2.CascadeClassifier("data/haarcascade_eye.xml")
recognizer = cv2.face.LBPHFaceRecognizer_create()
recognizer.read("trainner.yml")

labels = {}
with open("labels.pickle", "rb") as f:
    og_labels = pickle.load(f)
    labels = {v: k for k, v in og_labels.items()}

cap = cv2.VideoCapture(0)

while (True):
    # Que capture el frame-by-frame
    ret, frame = cap.read()
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    faces = face_cascade.detectMultiScale(gray, scaleFactor=1.3, minNeighbors=5)
    for (x, y, w, h) in faces:
        roi_gray = gray[y:y + h, x:x + w]
        roi_color = frame[y:y + h, x:x + w]

        # Para reconocer necesitamos un modelo de prediccion /keras /tensorflor /pytorch / sickit-learn
        id_, conf = recognizer.predict(roi_gray)
        if conf >= 45:  # and conf <= 85:
            print(id_)
            print(labels[id_])
            font = cv2.FONT_HERSHEY_SIMPLEX
            name = labels[id_]
            color = (255, 255, 255)
            stroke = 2
            cv2.putText(frame, name, (x, y), font, 1, color, stroke, cv2.LINE_AA)

        img_item = "my-image.png"
        cv2.imwrite(img_item, roi_gray)

        # Realizar el triangulo
        color = (255, 0, 0)  # El color esta en BGR
        grosor = 2
        end_cord_x = x + w
        end_cord_y = y + h
        cv2.rectangle(frame, (x, y), (end_cord_x, end_cord_y), color, grosor)
        eyes = eye_cascade.detectMultiScale(roi_gray)
        for (ex, ey, ew, eh) in eyes:
            cv2.rectangle(roi_color, (ex, ey), (ex + ew, ey + eh), (0, 255, 0), 2)

    # Muestre el frame resultado
    cv2.imshow("Frame", frame)

    # Si apretamos la Q en el frame, va a exitear
    if cv2.waitKey(20) & 0xFF == ord('q'):
        break

# Cuando este todo terminado, que lanze la captura
cap.release()
cv2.destroyAllWindows()
